bots/pr/src/main/java/org/openjdk/skara/bots/pr/ReviewCoverage.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.skara.bots.pr;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.openjdk.skara.forge.PullRequest;
+import org.openjdk.skara.forge.PullRequestUtils;
+import org.openjdk.skara.forge.Review;
+import org.openjdk.skara.vcs.Repository;
+import org.openjdk.skara.vcs.git.GitRepository;
+
+public class ReviewCoverage {
+
+    private final Logger log = Logger.getLogger("org.openjdk.skara.bots.pr");
+    private final boolean ignoreStaleReviews;
+    private final boolean includeSimpleMerges;
+    private final Repository repo;
+
+    public ReviewCoverage(boolean ignoreStaleReviews, boolean includeSimpleMerges, Repository repo) {
+        this.ignoreStaleReviews = ignoreStaleReviews;
+        this.includeSimpleMerges = includeSimpleMerges;
+        this.repo = repo;
+    }
+
+    public boolean covers(Review review, PullRequest pr) {
```

Comment:
In fact, during a single invocation of `CheckRun.checkStatus`, `ReviewCoverage.covers` is called **four** times for the same review and PR. The first time to identify if a reviewer should be added to the "Reviewed-by" list in a commit message. The second time to identify if a review should be labelled with "Re-review required" or "Review applies to ..." in the "Reviewers" bulleted list in the PR body. The third and fourth times to decide whether the "Reviewed-by" list in the commit message should be amended:

  1. CheckRun:1329
  2. CheckRun:1411
  3. CheckRun:1416 (x 2)

FYI, when I introduced trivial caching, the savings in my test scenario were very modest. Without any cache, on my machine, `org.openjdk.skara.bots.pr.CheckTests#acceptSimpleMerges` on average took 18,166 ms to complete. With the cache, on the same machine, that same test took 17,784 ms. This is 2% improvement.

I haven't tried to address your other concern for `PullRequestUtils::targetHash` yet.

Reply:
I tried to cache `targetHash`. The average time went down a bit more: now it's 17,627 ms. I'll push what I have, and you can see for yourself if you like it or not.

Reply:
I agree that it may not amount to much, especially not in a local test where the Git repository is minimal. I would expect some of these operations to be noticeably more expensive in properly sized repositories like the JDK. It's also about not suffering from the death of a thousand cuts. It's very easy for improvements like this to grow resource usage over time. So I appreciate you making the effort to implement caching. At least this way we know we aren't adding an unreasonable amount of extra work to a CheckRun.